### PR TITLE
Add theme support

### PR DESCRIPTION
**Pull request by safareli**
- update `Chart.init` to also take `Theme` as argument
- add `registerTheme` action
- update examples

### CODE REVIEW

File: src/ECharts/Chart.js [hunk 1]
```diff
@@ -1,8 +1,19 @@
...
+        return function() {
+            return echarts.init(el, theme);
+        };
+    };
+};
+
+exports.registerTheme = function(name) {
+    return function(theme) {
+        return function() {
+            echarts.registerTheme(name, theme);
```
Comment (safareli) [inline src/ECharts/Chart.js]: Does this check looks good?
```js
// if value is not `undefined` and is not of type `string` then it must be `Foreign` for which we only permit plain Objects.
if (theme !== undefined && typeof theme !== 'string' && !isObject(theme)) {
  throw new TypeError('Theme must be an Object')
}

function isObject(o) {
  return null != o && 
    typeof o === 'object' && 
    Object.prototype.toString.call(o) === '[object Object]';
}
```

File: example/src/Themes.js [hunk 2]
```diff
@@ -0,0 +1,132 @@
+exports.themeDark = (function() {
```
Comment (safareli) [inline example/src/Themes.js]: btw we can add module `ECharts.Theme` like this
```js
// by just requiring they are registered
require("echarts/theme/roma");
require("echarts/theme/dark");
// require rest of themes here
```
```purescript
dark :: Theme
dark = ByName "dark"
-- define rest of `Theme`s
```

Comment (cryogenian) [reply example/src/Themes.js]: :+1: Are you going to add it in this pr?

File: src/ECharts/Chart.js (hunk 1 above)
Comment (cryogenian) [reply src/ECharts/Chart.js]: Yeah, but please don't return `{}`

File: example/src/Bar.purs [hunk 3]
```diff
@@ -142,12 +142,12 @@ genInp = F.for (Arr.range 0 10) \i → do
   four ← random <#> ((_ + 0.3) >>> U.precise 2.0)
   pure {label, one, two, three, four}
 
-chart ∷ ∀ e. Eff (random ∷ RANDOM, dom ∷ DOM, echarts ∷ ET.ECHARTS, err ∷ EXCEPTION|e) Unit
-chart = do
+chart ∷ ∀ e. Maybe EC.Theme → Eff (random ∷ RANDOM, dom ∷ DOM, echarts ∷ ET.ECHARTS, err ∷ EXCEPTION|e) Unit
```
Comment (cryogenian) [inline example/src/Bar.purs]: Could you please leave `chart` signature as is and make it something like `chart = chart' Nothing`?

**Review by cryogenian — CHANGES_REQUESTED** [review]

File: example/src/Scatter.purs [hunk 4]
```diff
@@ -93,11 +94,17 @@ options sinData cosData = do
 
 chart ∷ ∀ e. Eff (dom ∷ DOM, echarts ∷ ET.ECHARTS, err ∷ EXCEPTION, random ∷ RANDOM|e) Unit
 chart = do
-  mbEl ← U.getElementById $ ElementId "scatter"
+  chart' (ElementId "scatter-1") Nothing
```
Comment (safareli) [inline example/src/Scatter.purs:97]: do you wanted something like this?

File: src/ECharts/Chart.js (hunk 1 above)
Comment (safareli) [reply src/ECharts/Chart.js]: do you want to also remove `return {}` from other places too?

File: src/ECharts/Theme.purs [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+  , infographic
+  , macarons
+  , roma
+  , shine
+  , vintage
+  ) where
+
+import Data.Foreign (Foreign)
+
+foreign import forceExport :: Foreign
```
Comment (safareli) [inline src/ECharts/Theme.purs:13]: did this, to make sure theme.js was included in output. Is non hacky  way to do it?

File: example/src/Scatter.purs (hunk 4 above)
Comment (cryogenian) [reply example/src/Scatter.purs:97]: Exactly :+1: This way it's backward compatible and we can make minor version update

File: src/ECharts/Theme.purs (hunk 5 above)
Comment (natefaubion) [reply src/ECharts/Theme.purs:13]: One option would be to omit the foreign requires, and just tell library consumers to make sure the themes are included in their builds.

**Review by cryogenian — CHANGES_REQUESTED** [review]

File: example/src/Scatter.purs (hunk 4 above)
Comment (safareli) [reply example/src/Scatter.purs:97]: why do we care about backward compatible updates in examples?

**Comment by cryogenian** [comment]
Oh! Yep!
Let's define an ADT
```purescript
data BuiltInTheme = Infographic | Macarons | Roma | Shine | Vintage | Dark 
Then define a func from it to string like

builtInThemeRoute :: BuiltInTheme -> String 
builtInThemeRoute = case _ of 
  Roma -> "echarts/theme/roma"
  _ -> ... 

builtinTheme :: BuiltInTheme -> Theme 
builtinTheme = case _ of 
  Roma -> "roma"
  _ -> ...
```
Then define a func that registers builtintheme
```purescript
foreign import registerBuiltin_ ::  String-> Eff _ Unit
registerBuiltin :: BuiltInTheme -> Eff _ Theme
registerBuiltIn bit = builtinTheme bit <$ registerBuiltin_ (builtinThemeRoute bit)
```
and in Theme.js
```javascript
exports.registerBuiltIn_ = function(route) {
  return function() {
    require(route);
  };
};
```
This way it's cleaner that registering builtin themes are effectful and they're actually predefined. And preserves notion of themes being open to extension by outside code.

**Comment by safareli** [comment]
@cryogenian the option looks nice!

But it might have issues with dead code elimination. I think code analyzers will not be able to determine required pathes which are dynamic, if that issue is not a problem Will implement it as you pointed out.

**Comment by natefaubion** [comment]
I would recommend not relying on dynamic requires.

**Comment by cryogenian** [comment]
@natefaubion In this case it's actually isn't `require` but just an effect that registers an object as theme. (Basically stores an object to global map). 
So, yeah I know that this will bundle all themes to output, I'm not sure that this is a problem though.

File: example/src/Scatter.purs (hunk 4 above)
Comment (cryogenian) [reply example/src/Scatter.purs:97]: That's not about examples, but about API. If we preserve `chart` signature, then API isn't changed and it's a minor version bump. 
I've just put comment here, not there :D

**Comment by natefaubion** [comment]
The bundler can't pick up opaque dynamic requires though.

File: example/src/Scatter.purs [hunk 6]
```diff
@@ -93,11 +94,17 @@ options sinData cosData = do
...
+  chart' (ElementId "scatter-3") (Just (ETheme.macarons))
+
+chart' ∷ ∀ e. ElementId → Maybe ETheme.Theme → Eff (dom ∷ DOM, echarts ∷ ET.ECHARTS, err ∷ EXCEPTION, random ∷ RANDOM|e) Unit
+chart' id theme = do
+  mbEl ← U.getElementById id
   case mbEl of
     Nothing → DT.traceAnyA "There is no element with scatter id"
     Just el → do
-      ch ← EC.init el
+      ch ← maybe EC.init EC.initWithTheme theme el
```
Comment (safareli) [inline example/src/Scatter.purs:107]: looks good?

**Comment by safareli** [comment]
I think it would be better to just state in readme to include themes client is want's to use in there bundler, as otherwise we have two options:

- Do what I have done currently. But, even if user is using just `Theme` type (no build in themes are sued), then `Theme` module will be bundled in output and I think, it will contain unused build in themes, which we don't want.

- Do as Maxim suggested, But bundlers will not be able to determine what should be included in output and clients might need to explicitly add theme pathes to there bundler configuration. 

Do you agree, or am I missing something?

**Review by cryogenian — APPROVED** [review]

**Comment by cryogenian** [comment]
Ok, I think this is good to go. BTW, we can also use one module per theme. This way if user import theme it's `require`d. But I think this is a matter of different pr.

**Pull request merged**